File: LRUCache.py
class DLL:
    def __init__(self, key=0, value=0):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None


class LRUCache:

    def __init__(self, capacity: int):
        self.head = DLL()
        self.tail = DLL()
        self.cache = {}
        self.n = capacity
        self.size = 0

        self.head.next = self.tail
        self.tail.prev = self.head

    def add_node(self, node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node
        return node

    def remove_node(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def pop_node(self):
        res = self.tail.prev
        self.remove_node(res)
        return res

    def reorder(self, node):
        self.remove_node(node)
        self.add_node(node)

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1

        node = self.cache.get(key)
        self.reorder(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        if key not in self.cache:
            if self.size < self.n:
                self.size += 1

            else:
                self.cache.pop(self.pop_node().key)

            self.cache[key] = self.add_node(DLL(key, value))

        else:
            node = self.cache.get(key)
            node.value = value
            self.reorder(node)

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: java_version/src/LRUCache.java
class DLLNode{
    public int key;
    public int value;
    public DLLNode next;
    public DLLNode prev;

    public DLLNode(int k, int v){
        this.key = k;
        this.value = v;
        this.next = null;
        this.prev = null;
    }
}

class LRUCache {
    public DLLNode head;
    public DLLNode tail;
    public Map<Integer, DLLNode> cache;
    public int size;
    public int n;

    public LRUCache(int capacity) {
        this.n = capacity;
        this.size = 0;
        this.head = new DLLNode(0, 0);
        this.tail = new DLLNode(0, 0);
        this.cache = new HashMap<Integer, DLLNode>();

        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    public void addNode(DLLNode node){
        node.next = this.head.next;
        node.prev = this.head;
        this.head.next.prev = node;
        this.head.next = node;
    }

    public void delNode(DLLNode node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    public DLLNode pop(){
        DLLNode node = this.tail.prev;
        this.delNode(node);
        return node;
    }

    public void reorder(DLLNode node){
        this.delNode(node);
        this.addNode(node);
    }

    public int get(int key) {
        if (!this.cache.containsKey(key))
            return -1;

        DLLNode node = this.cache.get(key);
        this.reorder(node);
        return node.value;

    }

    public void put(int key, int value) {
        if (!this.cache.containsKey(key)){
            if (this.size < this.n){
                this.size++;
            }
            else{
                DLLNode node = this.pop();
                this.cache.remove(node.key);
            }
            DLLNode node = new DLLNode(key, value);
            this.addNode(node);
            this.cache.put(key, node);
        }
        else{
            DLLNode node = this.cache.get(key);
            node.value = value;
            this.reorder(node);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */